utildb/gen_deleted_accounts.go

```diff
@@ -59,100 +67,280 @@ func readAccounts(ch chan proxy.ContractLiveliness, deleteHistory *map[common.Ad
+		deleteHistory = nil
+	}()
+
+	for {
+		select {
+		case <-abort.Wait():
+			return
+		case contract, ok := <-orderedResults:
+			{
+				if !ok {
+					return
+				}
+				des, res := readAccounts(contract.liveliness, &deleteHistory)
+				if len(des)+len(res) > 0 {
+					err := ddb.SetDestroyedAccounts(contract.tx.Block, contract.tx.Transaction, des, res)
+					if err != nil {
+						errChan <- err
+						return
+					}
+				}
+			}
 		}
+	}
+}
+
+// resultCollector collects results from workers in round-robin fashion and sends them to a single channel.
+func resultCollector(wg *sync.WaitGroup, cfg *utils.Config, workerOutputChannels map[int]chan txLivelinessResult, abort utils.Event) chan txLivelinessResult {
+	orderedResults := make(chan txLivelinessResult, cfg.Workers)
+	wg.Add(1)
+	go func() {
+		defer close(orderedResults)
+		defer wg.Done()
 
-		if tx.Transaction < utils.PseudoTx {
-			err = genDeletedAccountsTask(tx, processor, ddb, &deleteHistory, cfg)
-			if err != nil {
-				return err
+		// round-robin to collect results from workers
+		for {
+			for i := 0; i < cfg.Workers; i++ {
+				select {
+				case <-abort.Wait():
+					return
+				case res, ok := <-workerOutputChannels[i]:
+					if !ok {
+						return
+					}
+
+					// filter out txs with no liveliness actions
+					if res.liveliness != nil && len(res.liveliness) > 0 {
+						select {
+						case <-abort.Wait():
+							return
+						case orderedResults <- res:
+						}
+					}
+				}
 			}
+		}
+	}()
+	return orderedResults
+}
+
+// txProcessor launches workers to process transactions in parallel.
+func txProcessor(wg *sync.WaitGroup, cfg *utils.Config, workerInputChannels map[int]chan *substate.Transaction, processor executor.TxProcessor, abort utils.Event, errChan chan error, log logger.Logger) map[int]chan txLivelinessResult {
+	// channel for each worker to send results
+	workerOutputChannels := make(map[int]chan txLivelinessResult)
+	for i := 0; i < cfg.Workers; i++ {
+		workerOutputChannels[i] = make(chan txLivelinessResult)
+	}
+
+	for i := 0; i < cfg.Workers; i++ {
+		wg.Add(1)
+		go func(workerId int) {
+			defer func() {
+				close(workerOutputChannels[workerId])
+				wg.Done()
+			}()
+
+			for {
+				select {
+				case <-abort.Wait():
+					return
+				case tx, ok := <-workerInputChannels[workerId]:
+					if !ok {
+						return
+					}
+					// Process sorted transactions
+					livelinessArr, err := genDeletedAccountsTask(tx, processor, cfg, log)
+					if err != nil {
+						errChan <- err
+						return
+					}
+
+					select {
+					case <-abort.Wait():
+						return
+					case workerOutputChannels[workerId] <- txLivelinessResult{livelinessArr, tx}:
+					}
+				}
+			}
+		}(i)
+	}
+
+	return workerOutputChannels
+}
+
+// taskFeeder feeds tasks to workers in round-robin fashion.
+func taskFeeder(wg *sync.WaitGroup, cfg *utils.Config, iter substate.SubstateIterator, lastBlock uint64, abort utils.Event, log logger.Logger) map[int]chan *substate.Transaction {
```

Comment:
I dont really Like the chanels being placed in a map - since the key in the map is a number, you can use a normal array because its a ordered structure hence it serves the same purpose.

Reply:
Makes sense. Fixed.